Ignore Yarn Plug'n'Play installation dependencies

PatrickShaw opened the pull request:
Yarn Plug'n'Play (a relatively recent package management tool typically used for managing JavaScript dependencies) dependencies and some other files are committed to the repository if you're using [Zero Installs](https://yarnpkg.com/features/zero-installs) guidelines. As such, whatever code you've installed as a dependency (i.e. Code that's not yours) ends up slightly skewing Linguist's analysis statistics.

For example, my [monorepo](https://github.com/PatrickShaw/monorepo/tree/70b24446932fc7f68893f6f500d5ca208be43b12) uses almost exclusively TypeScript, however, ~most of the dependencies in `.yarn/cache/*`~ `.pnp.js` is a giant JavaScript file generated by yarn and so it looks like my monorepo is primarily made up of JavaScript.

This PR should fix this issue for anyone using Yarn Plug'n'Play.

<!--- Briefly describe your changes in the field above. -->

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- Feel free to remove whole sections, not points within the sections, that do not apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] **I am associating a language with a new file extension.**
  - [ ] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      - https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack
  - [ ] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am adding a new language.**
  - [ ] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      -  https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack
  - [ ] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  <!-- Update the Lightshow URLs below to show the grammar in action if you included one. -->
  - [ ] I have included a syntax highlighting grammar: https://github-lightshow.herokuapp.com/
  - [ ] I have updated the heuristics to distinguish my language from others using the same extension.

- [x] **I am fixing a misclassified language**
  - [x] I have included a new sample for the misclassified language:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am changing the source of a syntax highlighting grammar**
  <!-- Update the Lightshow URLs below to show the new and old grammars in action. -->
  - Old: https://github-lightshow.herokuapp.com/
  - New: https://github-lightshow.herokuapp.com/

- [ ] **I am updating a grammar submodule**
  <!-- That's not necessary, grammar submodules are updated automatically with each new release. -->

- [ ] **I am adding new or changing current functionality**
  <!-- This includes modifying the vendor, documentation, and generated lists. -->
  - [ ] I have added or updated the tests for the new or changed functionality.

- [ ] **I am changing the color associated with a language**
  <!-- Please ensure you have gathered agreement from the wider language community _before_ opening this PR -->
  - [ ] I have obtained agreement from the wider language community on this color change.
    - [URL to public discussion]
    - [Optional: URL to official branding guidelines for the language]

lildude commented:
Thanks for your PR, however it's completely unnecessary.

Absolutely _none_ of the files under `.yarn/cache/*` should be counted in your language stats because the `cache` path is already ignored at https://github.com/github/linguist/blob/309acde92e04c86466441afc12992274be938473/lib/linguist/vendor.yml#L12-L13

We can confirm this by analysing your repo on the command line:

```console
$ github-linguist . --breakdown
73.58%  JavaScript
26.20%  TypeScript
0.21%   CSS
0.01%   HTML

JavaScript:
.eslintrc.js
.pnp.js
babel.config.js
build-packages/gulp-staged/src/index.js
build-packages/gulp-staged/writeme.config.js
build-packages/gulp-status-git-filter/src/file.js
build-packages/gulp-status-git-filter/src/index.js
build-packages/gulp-status-git-filter/writeme.config.js
build-packages/monorepo-config/src/index.js
build-packages/stream-filter-glob/src/index.js
build-packages/tslint-preset-core/writeme.config.js
build-packages/tslint-preset-react/writeme.config.js
build-packages/winston-formats/src/index.js
byexample/cli/bin/index.js
byexample/storybook-addon-auto-knobs/.storybook/TestComponent.jsx
byexample/storybook-addon-auto-knobs/.storybook/addons.js
byexample/storybook-addon-auto-knobs/.storybook/config.js
byexample/storybook-addon-auto-knobs/.storybook/createStories.jsx
byexample/storybook-addon-auto-knobs/.storybook/webpack.config.js
byexample/storybook-knobified/register.js
faultjs/fault-benchmark/src/hooks/babel.config.js
faultjs/fault-benchmark/webpack.config.js
gulpBabelRegister.js
misc/convict-util/writeme.config.js
misc/hook-schema/examples/logged-read-json/index.js
misc/hook-schema/writeme.config.js
misc/logger/examples/file-and-console/index.js
misc/logger/writeme.config.js
misc/markdown-util/writeme.config.js
misc/sinon-stub-functions/test/exports-module.js
misc/writeme-core/examples/from-package-directory/index.js
misc/writeme-core/writeme.config.js
monorepo.config.js
patrick-shaw/eslint-config-core/test/test.js
patrick-shaw/eslint-config-core/writeme.config.js
patrick-shaw/eslint-config-plugin-cinder/writeme.config.js
patrick-shaw/eslint-config-plugin-import/writeme.config.js
patrick-shaw/eslint-config-plugin-react/test/test.js
patrick-shaw/eslint-config-plugin-react/writeme.config.js
patrick-shaw/eslint-config-plugin-typescript/test/test.js
patrick-shaw/eslint-config-plugin-typescript/writeme.config.js
patrick-shaw/eslint-config-preset/writeme.config.js
patrick-shaw/my-resume/webpack.config.js
prettier.config.js
test/helpers/globals.js
test/require/babel.js
writeme.config.js

TypeScript:
__mocks__/@storybook/addon-actions.ts
__mocks__/@storybook/addon-knobs.ts
[... truncated for brevity ...]
$
```

In fact, your proposed changes won't make the slightest difference as `.yarn/` doesn't appear anywhere in the analysis. 😁 

If we look at your analysis by bytes of code, as Linguist does, we see you have waaaaay more Javascript than TypeScript:

JavaScript 2.03 MB
TypeScript 741 KB
CSS 5.99 KB
HTML 297 Bytes

_One_ file is responsible for 2.01MB of the 2.03MB of JavaScript - this bad boy https://github.com/PatrickShaw/monorepo/blob/development/.pnp.js

That file appears to be generated by yarnpkg/pnp so if this PR were to do anything, it should probably be adding `.pnp.js` to the list of generated files.

If we manually mark it as generated with an override and perform another analysis, we get:

TypeScript 741 KB
JavaScript 21.2 KB
CSS 5.99 KB
HTML 297 Bytes

... as you can see in my fork of your repo in which I've made this change - https://github.com/lildude/monorepo

PatrickShaw commented:
*facepalm* Yep you're right. In fact, on top of what you mentioned, @lildude it didn't even occur to me that the dependencies in the cache are `.gz` files.

I've added `.pnp.js` instead (and it's .cjs equivalent + .mjs for the potential future)

lildude commented:
I think these should be marked as generated, not vendored, as the top of the file states it's generated:

```
      "This file is automatically generated. Do not touch it, or risk",
      "your modifications being lost. We also recommend you not to read",
      "it either without using the @yarnpkg/pnp package, as the data layout",
      "is entirely unspecified and WILL change from a version to another."
```

PatrickShaw commented:
Yep it's a generated file 👍
Moved it there and added a test. Not sure if it's supposed to go in the `test_glob.rb` like the comment says or `test_generated.rb` but I put t in `test_generated.rb`, lemme if that's wrong.

pchaigno commented on lib/linguist/generated.rb:
```diff
@@ -386,6 +387,13 @@ def npm_shrinkwrap_or_package_lock?
+    #
+    # Returns true or false.
+    def generated_yarn_plugnplay?
+      name.match(/(^|\/)\.pnp\.(c|m)?js$/)
```
This method is going to return [a `MatchData` object](https://ruby-doc.org/core-2.5.1/Regexp.html). Please add `!!` to return a boolean.

PatrickShaw replied:
Okay, done :+1: was copying 1 of the 2 other methods that didn't use `!!` - Refactored them too.
